### ISSUE TITLE
Error with VS2017 compile

### ISSUE THREAD

**Issue by kliffgomel** [jb-alvarado/media-autobuild_suite#736]
I compile ffmpeg with the support of NVIDIA CUDA SDK. Visual Studio 2017 is installed on the system, but an error occurs when make.

┌ ffmpeg git ........................................... [Updates found]
├ Changing options to comply to nonfree...
FFmpeg and related apps will depend on CUDA SDK!
FFmpeg and related apps will depend on Nvidia drivers!
├ Compiling static FFmpeg...
├ Running configure...
├ Running make...
Likely error:
                 from D:/ffmpeg-autobuild/build/ffmpeg-git/libavfilter/opencl.c:23:
C:/PROGRA~1/NVIDIA~2/CUDA/v9.1/include/CL/cl_platform.h:1274:26: warning: "_MSC_VER" is not defined, evaluates to 0 [-Wundef]
 #if defined( _WIN32) && (_MSC_VER >= 1500)
                          ^~~~~~~~
CC      libavfilter/opencl/overlay.o
CC      libavfilter/opencl/unsharp.o
NVCC    libavfilter/vf_scale_cuda.ptx
c:\program files\nvidia gpu computing toolkit\cuda\v9.1\include\crt/host_config.h(135): fatal error C1189: #error:  -- unsupported Microsoft Visual Studio version! Only the versions 2012, 2013, 2015 and 2017 are supported!
vf_scale_cuda.cu
make: *** [/build/ffmpeg-git/ffbuild/common.mak:95: libavfilter/vf_scale_cuda.ptx] Error 2
make failed. Check D:/ffmpeg-autobuild/build/ffmpeg-git/build-static-64bit/ab-suite.make.log
This is required for other packages, so this script will exit.

https://0x0.st/sbFO.zip

**Comment by wiiaboo** [comment]
> scale_cuda (needs CUDA SDK and MSVC **_2015_** installed and cl.exe in PATH)

**Comment by wiiaboo** [comment]
Also, there was an issue recently about the same thing, look around for it.

**Comment by kliffgomel** [comment]
CUDA SDK support VS 2017 and early I compile ffmpeg with this script and VS 2017. Problem arise when I updated CUDA SDK to version 9.1.

**Comment by wiiaboo** [comment]
Well, this isn't called Nvidia-autobuild-suite, so... I don't really care enough to install cuda and msvc again just to try to fix an issue unrelated to the suite.

**Comment by schmidthubert** [comment]
What version does `cl /?` actually show then? Maybe there's indeed something wrong with the _MSC_VER checking in host_config.h which has a time stamp from Nov 4th 2017 but we have Feb 2018 already now. Maybe that `#if _MSC_VER < 1600 || _MSC_VER > 1911` check can be patched to fix this error, otherwise tell the NV CUDA SDK dev guys about it  ...

**Comment by wiiaboo** [comment]
There, sort it out yourself: https://github.com/jb-alvarado/media-autobuild_suite/issues/732
tl;dr: if `which cl.exe` doesn't work in mintty and isn't pointing to msvc2015's cl.exe, you're not doing it right and it's not my fucking problem. Complain to Nvidia about it, not FFmpeg or especially not me.

**Comment by wiiaboo** [comment]
Any more issues about this and I'm removing any mention to scale_cuda in the suite.

**Comment by schmidthubert** [comment]
Well I don't have any problems , was just trying to help. And just go ahead then, this is git so someone else will pick up from here. Besides, it works with VS2013 as well, cheers! :+1:

**Comment by kliffgomel** [comment]
A simple question and unexpectedly sharp reaction from the developer ...
I saw the requirements for compiling cuda filters, but I personally do not understand why MSVC 2015 is declared there only? The CUDA SDK supports MSVC 2017 compilers and my problem was not that.
In my case, the problem arose after MSVC 2017 was upgraded to version 19.12.25835 (15.5.6), which had not yet been officially tested by CUDA SDK developers, and therefore was disabled in the list of supported compilers. Before the update, everything was perfectly compiled with the script from MSVC 2017.
To solve the problem, I had to install in MSVC 2017 a set of compilation tools of the previous version and run the script from the appropriate environment of the required compiler, as indicated in this article https://blogs.msdn.microsoft.com/vcblog/2017/11/15/side- by-side-minor-version-msvc-toolsets-in-visual-studio-2017.

**Comment by wiiaboo** [comment]
> A simple question and unexpectedly sharp reaction from the developer ...

Because I'm tired of replying with the same answers to something that's out of my control and that I frankly couldn't care less about. My patience for broken shit isn't limitless.

**Comment by schmidthubert** [comment]
So it is as I thought, 1912 > 1911 and the CUDA SDK 9.1 is older than VS2017 v1912 and didn't and doesn't know about anything >= 1911 yet, simple thing. Thanks for the heads up! :+1:

**Comment by wiiaboo** [comment]
I'd suggest someone tell Nvidia to stop doing retarded checks, or else it'll keep being broken everytime MS releases a minor update to VS2017.

**Comment by kliffgomel** [comment]
This may not be the right advice, but you can disable compiler version checking at your own risk even by fixing the host_config file.h in [INSALL_DIR]\NVIDIA GPU Computing Toolkit\CUDA\v9.1\include\crt\

**Comment by Baddest-Juju** [comment]
I appreciate all the work you've done on this project @wiiaboo, I probably would not have been able to get a proper Windows binary of ffmpeg compiled without your help. Thank you.

That said, getting all worked up over such a little thing doesn't accomplish anything. I too, admittedly, glossed over the fact 2015 was what was required. I tried many things before stumbling onto your project so my brain was pretty fried already. Compiling ffmpeg as a Windows binary is such a pain.

Just as an experiment, I did try to disable the version check in CUDA to see if either of the 2017 compilers would work. No dice. My suggestion is to make a new tag specifically for CUDA issues so we can more easily help ourselves, and to make expressly clear in your readme that the **VS2017 compilers do not work**, and to download the VC++ 2015.3 v140 toolset component from the Visual Studio Installer.

Here's the batch file I made for VS2017 x64, made to be put in the same directory as media-autobuild_suite.bat. The line 2 path will be slightly different for VS2017 Professional or Enterprise.

```
@setlocal enableextensions
@call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvarsall.bat" x64 -vcvars_ver=14.0
@set "PATH=%PATH%;C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64"
@start /b /i /d "%~dp0" media-autobuild_suite.bat
```

I'm tagging @valesho from the other issue so they see this as well.

And thank you again @wiiaboo for all your hard work. May your coding lead to less aggravation in the future. I suggest cinnamon tea, it's very soothing.

**Comment by wiiaboo** [comment]
Does setting vcvarsall.bat do anything at all since you're just appending msvc 2015 path right after?
And why not use https://github.com/jb-alvarado/media-autobuild_suite/commit/ca57dfab6349563838c6b577081b934390a05516

**Comment by Baddest-Juju** [comment]
>Does setting vcvarsall.bat do anything at all since you're just appending msvc 2015 path right after?

vcvarsall.bat didn't set the path when set to 14.0 like I hoped, so the next thing I did was just toss that in as a local variable. It hadn't occurred to me that vcvarsall.bat might then be useless, I suppose I was thinking CUDA might fuss. I'll go and test that out.

>And why not use ca57dfa

Didn't know about it - looks like it's very recent. I can't say I like it though. I'd much rather take care of it at the same directory level as the media-autobuild_suite batch.

Ideally I'd love to just be able to pass it into the batch file like this:

``media-autobuild_suite.bat -cl "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64"``

Then I could just embed it into a shortcut.

**Comment by wiiaboo** [comment]
https://github.com/jb-alvarado/media-autobuild_suite/commit/ca57dfab6349563838c6b577081b934390a05516 is even better because you can set it for different bitness.

**Comment by Baddest-Juju** [comment]
>ca57dfa is even better because you can set it for different bitness.

Eh, to be honest I have a hard time imagining someone looking to do this sort of really high-end CUDA encoding that runs an x86 operating system.

**Comment by Baddest-Juju** [comment]
Oh, looks like compiling finished. You were right, calling vcvarsall doesn't make a difference. It's not setting the path correctly and it doesn't affect the compiling at all.

**Comment by kliffgomel** [comment]
And still, the most correct thing is to set the environment variables for the VS compiler via the call to vcvarsall.bat, since this is the method recommended by Microsoft. When you install VS2017 environment variables are not specified at all, so to build a project it's not enough just to specify the path to the compiler.
Indeed, probably not so many users who compile a project with support for CUDA. Therefore, in my opinion, the best solution would be to allow users to set environment variables via vcvarsall.bat for a particular architecture and a set of compilation tools autonomously, and in media-autobuild_suite.bat add support for the installed environment variables as I suggested earlier in https://github.com/jb-alvarado/media-autobuild_suite/commit/f5ed7cf97d6fd3d591d4c916bead7d79bafc6a07.

**Comment by Baddest-Juju** [comment]
>And still, the most correct thing is to set the environment variables for the VS compiler via the call to vcvarsall.bat, since this is the method recommended by Microsoft.

Then Microsoft should fix their batch file, because that did not work.

**Comment by wiiaboo** [comment]
This isn't meant for building projects though, Nvidia's CUDA compiler just needs the cl.exe, nothing more.

**Comment by kliffgomel** [comment]
Personally, everything works fine for me, I start building of the project with my script:
call "c: \ Program Files (x86) \ Microsoft Visual Studio \ 2017 \ Community \ VC \ Auxiliary \ Build \ vcvarsall.bat" amd64 -vcvars_ver = 14.0
cd / d d: \ ffmpeg-autobuild
call media-autobuild_suite.bat
But you have to manually delete / I when calling mintty from media-autobuild_suite.bat.

**Comment by kliffgomel** [comment]
And once again, in VS2017 when compiling with a set of tools from VS2015, specify in environment variables only the path to the compiler is not enough.

**Comment by Baddest-Juju** [comment]
>But you have to manually delete / I when calling mintty from media-autobuild_suite.bat.

I prefer the method where I don't have to redo edits whenever the project updates.

Adding 2015's cl.exe to PATH works fine. I already built ffmpeg w/ CUDA support using this method without calling vcvarsall.bat.

**Comment by wiiaboo** [comment]
Then I'll await another PR where you don't break everything else just for the sake of using vsvarsall.

**Comment by kliffgomel** [comment]
I do not understand what can break if to delete / I? It only will allow to set temporarily necessary variable environments and all.
As for the compiler, I added manually a way to cl.exe to the last update too, and then it ceased to work.
And all this it is wrong to write in global variables on a constant basis a way to the compiler for specific architecture-.
However, everyone can select.

**Comment by Baddest-Juju** [comment]
@wiiaboo Give this a spin.

``for /f "tokens=2*" %%a in ('reg query "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\VisualStudio\VC\19.0\x64\x64" /v Compiler 2^>^&1^|find "REG_"') do @set "PATH=%PATH%;%%b``

**Comment by wiiaboo** [comment]
@kliffgomel at least one breakage https://github.com/jb-alvarado/media-autobuild_suite/issues/752
Most certainly others since there's a lot of scripts involved in this.

**Comment by wiiaboo** [comment]
@Baddest-Juju I don't have that "\VC" sub-key. And the registry or VSxxxCOMNTOOLs has been deprecated for vswhere.exe.

**Comment by wiiaboo** [comment]
Well, fuck this shit, tbh. It needs people to manually alter headers to fix Nvidias retarded MSC_VER check, since they're installed to ProgramFiles, which sed can't access.
I'll just one more check and I hope that's the end.

**Comment by Baddest-Juju** [comment]
>I don't have that "\VC" sub-key [it's in WOW6432Node].

Yes, that's where it is for me too. WOW6432Node is a compatibility layer, it should redirect there automatically.

**Comment by Baddest-Juju** [comment]
@wiiaboo Me again, hello.

I tested the code I recommended to you prior by editing the latest script to read like this:

```
get_cl_path() {
    local vswhere="$(cygpath -u "$(cygpath -F 0x002a)/Microsoft Visual Studio/Installer/vswhere.exe")"
    if [[ -f "$vswhere" ]]; then
        local clpath
        local regpath="/HKLM/Software/Microsoft/VisualStudio/VC/19.0"
        if [[ $bits = 32bit ]]; then
            clpath="$(regtool -qW get "$regpath/x86/x86/Compiler")"
        elif [[ $bits = 64bit ]]; then
            clpath="$(regtool -qW get "$regpath/x64/x64/Compiler")"
        fi
        [[ -z "$clpath" ]] && return 1
        clpath="$(dirname "$(cygpath -u "$clpath")")"
        [[ ! -f "$clpath"/cl.exe ]] && return 1
        export PATH="$clpath":$PATH
    fi
}
```

As you can see, I removed the first method and just used this method you put in as a backup.

Here's a copy of my registry:

![image](https://user-images.githubusercontent.com/33403660/36497266-6e01eb80-1700-11e8-81f8-ea84627f23f7.png)
![image](https://user-images.githubusercontent.com/33403660/36497210-3f550402-1700-11e8-91d0-ce481544ca92.png)

As you can see, the registry entries only exist in WOW6432Node, and yet your code finds them. This is working as intended, WOW6432Node is a compatibility layer for Windows and the script is being automatically redirected to it without you having to do a thing.

I do recommend you make this the main method instead of the backup method, as this is compatible with both VS2015 and VS2017 w/ the 2015 Toolset.

**Comment by wiiaboo** [comment]
No, vswhere is the officially (by Microsoft) recommended way to get information of Visual Studio installations. Magic registry keys is deprecated.

**Comment by Baddest-Juju** [comment]
? But I installed this copy of VS2017 just last week. Strange.

Well.. either way seems to work, so hooray?

**Comment by wiiaboo** [comment]
Of course it works, because you left the check for vswhere, while removing everything that used it, for some reason.

**Comment by Baddest-Juju** [comment]
Perhaps I misunderstood the code? I thought the first method was the primary method you went with for retrieving cl.exe, while the apparently deprecated registry method was there as a backup. You stated in the commit note it was untested, so I wanted to test it for you.

**Comment by wiiaboo** [comment]
Untested for actual MSVC2015 installations, since that's impossible to do with community.

**Comment by Baddest-Juju** [comment]
Well.. for what it's worth, it did compile correctly. Sorry to bother you.
